Chosen combined solution by Group 8

muzaluisa opened the pull request:
First version to be refactored by everyone :)
- Movement class is added
- Yarinka's trick is added
- Complexity comments are added

muzaluisa committed:
Chosen combined solution by Group 8 

First version to be refactored by everyone :)

muzaluisa committed:
Minor typo fixes

muzaluisa committed:
Took into accounts Li comments on the previous ver

https://github.com/muzaluisa/codeU/pull/9#discussion_r129368376
Particularly:
- making functions private
- movement class
- typos and unused numpy library

luxiya01 commented on RearrangeCarGroup8Movements.py:
```diff
@@ -0,0 +1,195 @@
+
+    def print_log_file(self):
+        move_str = "Move car #%s from %s to %s" %(self.car, self.from_index, self.to_index)
+        logging.info(move_str)
```
Perhaps we don't really need this function? If we want to log the info, we can perhaps do `logging.info(movement_object.__str__())`?

luxiya01 commented on RearrangeCarGroup8Movements.py:
```diff
@@ -0,0 +1,195 @@
+        else:
+            movement.print_log_file()
+
+        self._move_car_to_empty_slot(car_to_move_position)
```
If we are to return a list of movement objects instead of printing right away, then perhaps we can remove line 151 - 154 and then just return `Movement(car_to_move, car_to_move_position, self.empty_position)` ^^

luxiya01 commented on RearrangeCarGroup8Movements.py:
```diff
@@ -0,0 +1,195 @@
+
+        while self.not_arranged_cars:
+            self._rearrange_one_car() # O(1)
+
```
What about moving `self._reset_variables_to_init_state()` to the very beginning of this function, so that we it automatically reset all states before checking not arranged cars? (Then we can remove the first `if` and `elif` checks. Since if the init state is equal to the final state, there will be no cars in `self.not_arranged_cars` and en empty list will be returned.) 

And then since we plan to return a movement object from `self._rearrange_one_car()`, we can let this function return a list of movements, and write another function that does the printing of the list?

luxiya01 commented on RearrangeCarGroup8Movements.py:
```diff
@@ -0,0 +1,195 @@
+        logging.info(self.initial_car_order)
+        logging.info('Final car order:')
+        logging.info(self.final_car_order)
+        logging.info('Moves:')
```
I have never used `logging` before, so I am a bit confused... When I was testing from my computer, I couldn't really find the log file, but I saw all the logging info being printed out to the console...

muzaluisa replied:
Should work now, I added logging.basicConfig(filename='RearrangeCarGroup8.log', filemode='w', level=logging.DEBUG)

luxiya01 reviewed (commented):
A few comments regarding the movement class and the log file XD

muzaluisa committed:
printing, logging updates

marijja commented on RearrangeCarGroup8Movements.py:
```diff
@@ -0,0 +1,206 @@
+        self.print_console = True
+        self._variables_to_init_state()
+
+    def print_moves_log_file(self, moves_list):
```
I think we should move printing functions to a new class to separate handling it from the logic (better separation of concerns! :D)

marijja commented on RearrangeCarGroup8Movements.py:
```diff
@@ -0,0 +1,206 @@
+        
+        info_message = 'Number moves is {0} > 1000, logging them..'.format(len(moves_list))
+        print info_message
+        logging.info('Initial car order:')
```
I'm wondering about putting this in a one function with newlines, as that's quite a lot of repetitive calls

muzaluisa replied:
```
logging.info('Initial car order:', self.initial_car_order)
logging.info('Final car order:', self.final_car_order)
```
Does this look good? or what you would recommend?

marijja commented on RearrangeCarGroup8Movements.py:
```diff
@@ -0,0 +1,206 @@
+            None
+        '''
+        
+        print 'Initial car order:', self.initial_car_order
```
We have here the same string to print as in the logging, one string to share between them would be useful

marijja commented on RearrangeCarGroup8Movements.py:
```diff
@@ -0,0 +1,206 @@
+        '''
+        
+        print 'Initial car order:', self.initial_car_order
+        print 'Final car order:', self.final_car_order
```
We could use some neat formatting (see f strings), but that applies only to Python 3 code

muzaluisa replied:
@marijja Ok, I try to change at least this function to format, that is available in 2.7 then?

marijja commented on RearrangeCarGroup8Movements.py:
```diff
@@ -0,0 +1,206 @@
+       self.current_car_order.
+       - Moves are either printed to the console or logged
+       - Stores and updates min_num_steps_left that shows min number of moves left
+        in order to avoid comparison current and final order lists that takes O(n)
```
We should give an overall complexity here and describe the number of moves we make

muzaluisa replied:
yes I will write that)

muzaluisa replied:
@marijja Could you please add tests, if you can?

muzaluisa replied:
This part is done I suppose)

marijja commented on RearrangeCarGroup8Movements.py:
```diff
@@ -0,0 +1,206 @@
+        '''
+        
+        if self.initial_car_order == self.final_car_order:
+            assert ValueError, 'Cars are already in the final order, no moves are required'
```
Already arranged cars is a desired situation, I don't think we should raise an error (I suppose this is raising, as the syntax feels unfamiliar to me - would be great if anyone explained)

muzaluisa replied:
Good point, the assert is not needed here probably, I fix that)

luxiya01 replied:
+1 for removing the assert. I think we can get rid of the checking entirely :)

marijja commented on testRearrangeCarGroup8.py:
```diff
@@ -0,0 +1,72 @@
+        self.final_order = [3, 1, 2, 0]
+        self.rc = RearrangeCars(self.init_order, self.final_order)
+
+    def test_rearrange_all_cars(self):
```
Maybe rearrange_sample_cars? (Minor change, but you know immediately what's up :D)

muzaluisa replied:
Sounds good!:)

marijja commented on testRearrangeCarGroup8.py:
```diff
@@ -0,0 +1,72 @@
+    def test_long_parking_lot(self):
+
+        '''Runs 30 test case with different initial and final orders
+        of length N to check that the final arrangement is correct
```
num_spots instead of N or reverse :D

marijja reviewed (commented):
I think we should add some tests (for example trivial cases with empty list or something like that), maybe basing on other solutions from our group

muzaluisa committed:
fixed logging into the file

and added complexity explanation

luxiya01 commented on testRearrangeCarGroup8.py:
```diff
@@ -0,0 +1,72 @@
+        '''
+
+        rc_same = RearrangeCars(self.final_order, self.final_order)
+        self.assertRaises(rc_same.rearrange_all_cars())
```
It's nice to test against already ordered input! But since we are planning to remove the assert from the `rearrange_all_cars()`, we can probably change line 44 to `self.assertEqual(self.rc.current_car_order, self.final_order)` ^^

luxiya01 commented on RearrangeCarGroup8Movements.py:
```diff
@@ -0,0 +1,206 @@
+        else:
+            self.print_moves_console(moves)
+            
+        self._variables_to_init_state()    
```
`self._variables_to_init_state()` (line 205) should probably be placed at the beginning of the function `rearrange_all_cars()`, otherwise I supposed all tests will fail since we will then set `current_car_order` to `initial_car_order` XD

muzaluisa replied:
True, fixed that 👍

luxiya01 commented on RearrangeCarGroup8Movements.py:
```diff
@@ -0,0 +1,206 @@
+        if len(moves) > 1000:
+            self.print_moves_log_file(moves)
+        else:
+            self.print_moves_console(moves)
```
What about making this function returning the moves, and then make another function that takes a list of movements as input parameter and does the actual printing? This isn't really what the assignment asked for, but will allow the result of rearrange_all_cars() to be integrated into larger modules XD

muzaluisa committed:
Updated print formatting, no assert 

Took into account some comments made yesterday night by Li and Marija

muzaluisa committed:
minor changes in test_already_proper_order

from assertRaises to assertEqual

muzaluisa committed:
test_rearrange_all_cars -> test_rearrange_sample_cars

luxiya01 committed:
Add test for empty parking lot (idea from @yarynapavlyshyn)

N.B. empty parking lot test fails right now

luxiya01 committed:
Add test for empty parking lot of length 1

Test passed

luxiya01 committed:
Remove variable self.empty_position to pass test_empty_parking_lot

muzaluisa committed:
fixed test_long_parking_lot comment typo

muzaluisa committed:
minor changes in syntax suggested by pylint ..

marijja committed:
Corrected printing

blmussati commented on RearrangeCarGroup8Movements.py:
```diff
@@ -0,0 +1,212 @@
+and print a series of moves from position x to y
+
+For our example, the series of moves should be e.g.:
+    move 1 to 2
```
Maybe change this to the actual string we are outputting, ie:
"Move car # 1 from 0 to 1", for consistency

luxiya01 replied:
Good idea! Will fix XD

luxiya01 committed:
Modify printing class and add comments

blmussati commented on RearrangeCarGroup8Movements.py:
```diff
@@ -0,0 +1,212 @@
+
+    def print_moves(self, moves):
+        if len(moves) > 1000:
+            self.print_moves_log_file(moves)
```
I think we should keep the comment of while we are doing the log and how it improves the code!
If I'm not mistaken, it should be on @muzaluisa previous version.

blmussati commented on RearrangeCarGroup8Movements.py:
```diff
@@ -0,0 +1,212 @@
+                'Final car order: {final_order}'.format(final_order=final_car_order) +
+                'Moves:')
+
+    def print_moves(self, moves):
```
I think that in the code we should also deal with the case that the initial_car_order is already the same of the final_car_order and hence no moves have to be done.

luxiya01 replied:
This seems to be handled by the set `self.not_arranged_cars)`. If the two lists are equal, then it wouldn't enter the while-loop and should just return XD

blmussati replied:
Ok thanks @luxiya01  :D

blmussati commented on testRearrangeCarGroup8.py:
```diff
@@ -0,0 +1,94 @@
+import numpy as np
+from RearrangeCarGroup8Movements import RearrangeCars
+
+class TestRearrangeCars(unittest.TestCase):
```
I think we should also test the case in which init_order and final_order are of different length - how our function handle that case? I believe it should throw an error of the sort "These are not the same car parks!" or "There is a mismatching in the car parks' sizes, please check them again". Something like that.

luxiya01 replied:
@blmussati I think this sounds like a very good idea! Do you think putting the check in the `__init__` function will be a good idea? Or should we put it in the method `rearrange_all_cars()`?

luxiya01 replied:
I temporarily put it in `rearrange_all_cars()`, but feel free to change it! :)

blmussati replied:
I cannot see the update but it might be because my git is still connecting to an old version, I'm not sure :/

blmussati replied:
I would putit in __init__ but since the deadline is passed I think rearrange_all_cars() will do :D

blmussati commented on testRearrangeCarGroup8.py:
```diff
@@ -0,0 +1,94 @@
+        '''Basic setting that will be used in several tests
+        '''
+        
+        self.init_order = [1, 2, 0, 3]
```
Should we name them initial_car_order and final_car_order for consistency through the files?

blmussati commented on RearrangeCarGroup8Movements.py:
```diff
@@ -0,0 +1,212 @@
+       - Moves are either printed to the console or logged
+       - Print moves in O(n), where n is the number of cars.
+       - More precisely at most n + k moves are taken, where k << n is the amount of movements
+       - where car is placed not to the final place
```
Don't understand what you mean with "the amount of movements where car is placed not to the final place"

luxiya01 replied:
I think the last line shall continue from the second-last line, will fix!

blmussati replied:
Great!

blmussati commented on RearrangeCarGroup8Movements.py:
```diff
@@ -0,0 +1,212 @@
+        not arranged cars
+
+        Storing car_position_dict helps to search indices of car in O(1) instead of O(n)
+        A set of not_arranged_cars helps to now when to stop in O(1) as well as to draw
```
helps to **know**, it's just a minor typo though :)

blmussati commented on RearrangeCarGroup8Movements.py:
```diff
@@ -0,0 +1,212 @@
+        '''Finds first car index not yet standing on the final destination
+        '''
+
+        car = random.choice(tuple(self.not_arranged_cars))  # O(1)
```
Why do we need to use random? Why we create a tuple?

muzaluisa replied:
I googled how to pick up in 0(1) the element from the set, this was the proposed solution. Since tuple are immutable I suppose we add tuple operation :)

blmussati replied:
Cool!

blmussati commented on testRearrangeCarGroup8.py:
```diff
@@ -0,0 +1,94 @@
+        self.rc.rearrange_all_cars()
+        self.assertEqual(self.rc.current_car_order, self.final_order)
+
+    def test_already_proper_order(self):
```
Sorry, didn't see this before :)

blmussati commented on testRearrangeCarGroup8.py:
```diff
@@ -0,0 +1,94 @@
+        rc_same_empty_pos.rearrange_all_cars()
+        self.assertEqual(rc_same_empty_pos.current_car_order, final_order_same_empty)
+
+    def test_long_parking_lot(self):
```
So this is the testing for when moves > 1000?

blmussati commented on testRearrangeCarGroup8.py:
```diff
@@ -0,0 +1,94 @@
+from RearrangeCarGroup8Movements import RearrangeCars
+
+class TestRearrangeCars(unittest.TestCase):
+
```
Another test worth doing is if one or both inputs given is/are the empty list!
How should we handle such case? If they are both empty, I would just raise the message "No moves required" or something like that. If just one is empty, we should raise an error, I believe.

luxiya01 replied:
I think that's handled by `test_empty_parking_lot` XD

blmussati replied:
Ok! I added a comment there now then, for clarification of terminology :)

blmussati reviewed (commented):
Hi everyone! I left some comments, mainly regarding the testing that we are missing and two questions on things which weren't clear to me. I fixed a typo in a comment, but it's a minor thing.

Please if someone can give a look!

luxiya01 committed:
Add test to check that initial_order is not modified after rearranging cars

luxiya01 committed:
Modify code to past the newly added test

self.current_car_position needs to be a copy of self.initial_car_position, they previously refered to the same object

luxiya01 committed:
Modifications according to comments from @blmussati

luxiya01 committed:
Change variable names for consistency

Based on comments from @blmussati

muzaluisa committed:
made printing to console for all cases, except long_parking_lot

muzaluisa committed:
remove unnecessary to exercise 8 file

muzaluisa committed:
remove unnecessary codeU2.py file

muzaluisa committed:
remove unnecessary codeU3.py file

muzaluisa committed:
Delete Assignment1-Gettingstarted.pdf

muzaluisa committed:
Update README.md

muzaluisa committed:
Rename RearrangeCarGroup8Movements.py to RearrangeCarGroup8.py

muzaluisa committed:
edit: from RearrangeCarGroup8 import RearrangeCars

luxiya01 committed:
Correct movement printing

blmussati commented on testRearrangeCarGroup8.py:
```diff
@@ -0,0 +1,118 @@
+
+    def test_empty_parking_lot(self):
+
+        '''Test the case when the parking lot is empty,
```
Just for reference, it's a car park, a parking lot is a slot of the car park. I think this test is for when there are no cars parked, am I right?
Should we change the name to clarify?

carmenmitru replied:
Hello @muzaluisa I've seent at testing that you use  unittest.main(verbosity=2).What is the purpose of this parameter?

yarynapavlyshyn replied:
@carmenmitru, Hi Carmen, here is the information I found.
The parameter is a verbosity level. There are 3 different levels:
- 0 (quiet): you just get the total numbers of tests executed and the global result
- 1 (default): you get the same plus a dot for every successful test or a F for every failure
- 2 (verbose): you get the help string of every test and the result

blmussati reviewed (commented):
Sorry, I didn't spot this yesterday! :( anyway the test work, it was just a matter of correct terminology

yarynapavlyshyn commented on testRearrangeCarGroup8.py:
```diff
@@ -0,0 +1,118 @@
+        
+        self.initial_car_order = [1, 2, 0, 3]
+        self.final_car_order = [3, 1, 2, 0]
+        self.rc = RearrangeCars(list(self.initial_car_order), list(self.final_car_order))
```
I do not understand why we need here to convert the list to list again.
May we just write 
self.rc = RearrangeCars(self.initial_car_order, self.final_car_order) ?

luxiya01 replied:
I did it here because in the test for initial_car_order, we checked self.rc.initial_car_order against self.car_order. If we don't list it, maybe these two variables will point to the same object?

muzaluisa replied:
@luxiya01 

Also, I guess that we need to test 
instead 
`self.assertEqual(rc_length_one.current_car_order, rc_length_one.current_car_order)`
this: 
`self.assertEqual(rc_length_one.current_car_order, [0])`
since we want to check if we have final order we want instead of final order that was saved in the class variable (that could be theorically initialized in a wrong way) :)
So should be smth like this:
 ```
def test_parking_lot_length_one(self):

        '''Test the case when the parking lot is of
        length one and there is no car placed in it
        '''
        
        rc_length_one = RearrangeCars([0], [0])
        rc_length_one.rearrange_all_cars(print_console=True)
        self.assertEqual(rc_length_one.current_car_order, [0])
```

luxiya01 replied:
@muzaluisa Hmm... Or perhaps we can add a test case that checks whether `rc.final_car_order` is the same as `self.final_car_order` after one call of `rearrance_all_cars`?

muzaluisa replied:
@luxiya01 Do you afraid that is self.initial_car_order changes, then the value of the object changes? 
This will not happen e.g. I tested as shown below :)

```
def setUp(self):

        '''Basic setting that will be used in several tests
        '''
        
        self.initial_car_order = [1, 2, 0, 3]
        self.final_car_order = [3, 1, 2, 0]
        self.rc = RearrangeCars(self.initial_car_order, self.final_car_order)
        
    def test_change_init_order(self):
       print 'change the reference variable'
       self.initial_car_order = [3, 1, 2, 0]
       self.assertEqual(self.rc.initial_car_order, [1, 2, 0, 3])
```

luxiya01 replied:
Yes, that was what I was afraid of, because I think it happened before... Previously, when the code in `RearrangeCarGroup8`contains the line `self.current_car_order = self.initial_car_order`, instead of `self.current_car_order = list(self.initial_car_order)`, these two variables refer to the same object and `self.intial_car_order` was modified. However, when I checked whether `self.rc.initial_car_order` was the same as `self.initial_car_order`, the test passed like a charm. When I printed them out, however, both had changed and became the same as `self.rc.current_car_order`, since they all pointed to the same object. That was why I listed them... OMG, hope this makes sense XD

luxiya01 replied:
@muzaluisa @yarynapavlyshyn So to sum it all up... I listed them so that if these variables were modified by calling out `rearrange_all_cars`, we will be able to detect them. i.e. without the listing, the `test_initial_order_preserved` wouldn't make much sense, since the variables that we are asserting equal to will be the same variable...

luxiya01 replied:
If, however, we don't intend to check whether `self.initial_car_order` and `self.final_car_order` are modified in the production code, then there is probably no need to list stuffs XD

muzaluisa replied:
@luxiya01 Now I got your explanation, many thanks. Basically indeed we need the line self.current_car_order = list(self.initial_car_order) in _variables_to_init_state(self). My tests also confirmed that. 

```
def test_initial_order_preserved(self):

        '''Tests whether, after having called rearrange_all_cars() once,
        the field self.initial_car_order is still unchanged

        Note: This test must pass for test_rearrange_sample_cars_twice to
        be meaningful
        '''
        
        print 'test for changing the var'
        
        self.initial_car_order = [3, 1, 2, 0]
        self.rc.rearrange_all_cars(print_console=True)
        print 'changed car order', self.rc.initial_car_order
        self.assertEqual(self.rc.initial_car_order, self.initial_car_order)
```

yarynapavlyshyn commented on testRearrangeCarGroup8.py:
```diff
@@ -0,0 +1,118 @@
+        should work without raising an error
+        '''
+        
+        rc_empty = RearrangeCars(list(), list())
```
And here it is better to pass just two "[]" parameters as list creation with "[", "]" is significantly faster than by list() and also it will be clearer what order we have and what are willing to obtain.

muzaluisa replied:
I will edit that, thanks :)

yarynapavlyshyn reviewed (commented):
Beautiful, great modularized, easy to read :)

muzaluisa committed:
fixed list() to [] and some minor assertEqual issue

added self.assertEqual(rc_length_one.current_car_order, [0])
to compare with the desired result instead of rc_length_one.final_car_order and 
similarly in self.assertEqual(rc_empty.current_car_order, [])
